core/src/main/scala/java/util/Currency.scala

```diff
@@ -0,0 +1,76 @@
+package java.util
+
+import locales.LocaleRegistry
+import scala.collection.{Map => SMap, Set => SSet}
+import locales.cldr.CurrencyDataInfo
+import locales.cldr.data.currencydata
+
+object Currency {
+  private val countryCodeToCurrencyCodeMap: SMap[String, String] = currencydata.regions.map{ r =>
+    r.countryCode -> r.currencies.find{ _.to.isEmpty }.map{ _.currencyCode }.get
+  }.toMap
+
+  private val all: SSet[Currency] = currencydata.infos.map{ info: CurrencyDataInfo =>
+    Currency(info.currencyCode, info.numericCode, info.digits, None)
+  }.toSet
+
+  private val currencyCodeMap: SMap[String, Currency] = all.toSeq.groupBy{ _.getCurrencyCode }.map{
+    case (currencyCode: String, matches: Seq[Currency]) => currencyCode -> matches.head
+  }
+
+  def getAvailableCurrencies(): SSet[Currency] = all
+
+  // NullPointerException - if locale or its country code is null
+  // IllegalArgumentException - if the country of the given locale is not a supported ISO 3166 country code.
+  def getInstance(locale: Locale): Currency = {
+    if (locale.getCountry == null || locale.getCountry.isEmpty) throw new NullPointerException
+
+    countryCodeToCurrencyCodeMap.get(locale.getCountry).flatMap{ currencyCodeMap.get }.getOrElse(throw new IllegalArgumentException)
+  }
+
+  def getInstance(currencyCode: String): Currency = currencyCodeMap(currencyCode)
+}
+
+final case class Currency private (currencyCode: String,
+    numericCode: Int,
+    fractionDigits: Int,
+    currencyLocale: Option[Locale] = None) {
+  def defaultLocale: Locale = currencyLocale.getOrElse(Locale.getDefault)
+
+  // Gets the ISO 4217 currency code of this currency.
+  def getCurrencyCode(): String = currencyCode
+
+  // Gets the default number of fraction digits used with this currency.
+  def getDefaultFractionDigits(): Int = fractionDigits
+
+  // Gets the name that is suitable for displaying this currency for the default DISPLAY locale.
+  def getDisplayName(): String = getDisplayName(defaultLocale)
+
+  // Gets the name that is suitable for displaying this currency for the specified locale.
+  def getDisplayName(locale: Locale): String = {
+    LocaleRegistry.ldml(locale).flatMap { ldml =>
+      ldml.getNumberCurrency(currencyCode).flatMap { _.displayNames.find{ _.count.isEmpty }.map{ _.name } }
+    }.getOrElse(currencyCode)
```

Comment:
This is the kind of thing that is always troubling because the Java spec is unclear on what to return if e.g. you don't have the display name. The only certain way to know is to test against the JVM